### ISSUE TITLE
bug report: 

### ISSUE THREAD

**Issue by zelogik** [docker-mailserver/docker-mailserver#3396]
### 📝 Preliminary Checks

- [X] I checked that all ports are open and not blocked by my ISP / hosting provider.
- [X] I know that SSL errors are likely the result of a wrong setup on the user side and not caused by DMS itself. I'm confident my setup is correct.
- [x] I searched the issue tracker but was unable to find my issue.
- [X] I have read:  
  - The [extended documentation in general](https://docker-mailserver.github.io/docker-mailserver/latest/) but found nothing to resolve the issue;
  - The [documentation on debugging](https://docker-mailserver.github.io/docker-mailserver/latest/config/debugging/), tried the proposed steps to debug the problem, but was still unable to resolve the issue;
  - This project's [Code of Conduct](https://github.com/docker-mailserver/docker-mailserver/blob/master/CODE_OF_CONDUCT.md) and I agree;
  - The [documentation about filing a bug report](https://docker-mailserver.github.io/docker-mailserver/latest/contributing/issues-and-pull-requests/#filing-a-bug-report).

### 👀 What Happened?

Pod is in "waiting" state ( see log in revelant output).

So dms don´t start.

### 👟 Reproduction Steps

Hi, 

Installation:
k8s baremetal + debian 11/12 + cri-o 

deploy:
kubectl apply -f dms.yml

error:
kubectl get events -w

tested:
- previous dms version, i suspect there is a problem with default apparmor profile docker/containerd VS cri-o but don´t find google/stackoverflow/github problems...
- change container.apparmor.security.beta.kubernetes.io/mailserver: [unconfined, runtime/default ], seem like the annotation is not applied.

simple setup like the below link, work without any issues.
https://kubernetes.io/docs/tutorials/security/apparmor/

I know that k8s is not officially supported, but at this point I don´t know if the problem is k8s, crio, dms, as I don´t have any issue with my k8s setup... and after struggling for many hours, i'm in despair...

ie: I know that there is a missing crt/key volumeMount (it's revelant?)

### 🐋 DMS Version

v12.1.0 (latest/v12.0/v11.3.1...)

### 💻 Operating System and Architecture

Debian 11/12, k8s v1.27 + cri-o 1.24.5

### ⚙️ Container configuration files

```yml
---
apiVersion: v1
kind: ConfigMap

metadata:
  name: mailserver.environment

immutable: false

data:
  TLS_LEVEL: modern
  POSTSCREEN_ACTION: drop
  OVERRIDE_HOSTNAME: xxxxxx
  FAIL2BAN_BLOCKTYPE: drop
  POSTMASTER_ADDRESS: xxxxxx
  UPDATE_CHECK_INTERVAL: 10d
  POSTFIX_INET_PROTOCOLS: ipv4
  ONE_DIR: '1'
  ENABLE_CLAMAV: '1'
  ENABLE_POSTGREY: '0'
  ENABLE_FAIL2BAN: '1'
  AMAVIS_LOGLEVEL: '-1'
  SPOOF_PROTECTION: '1'
  MOVE_SPAM_TO_JUNK: '1'
  ENABLE_UPDATE_CHECK: '1'
  ENABLE_SPAMASSASSIN: '1'
  SUPERVISOR_LOGLEVEL: warn
  SPAMASSASSIN_SPAM_TO_INBOX: '1'

  SSL_TYPE: manual
  SSL_CERT_PATH: /secrets/ssl/rsa/tls.crt
  SSL_KEY_PATH: /secrets/ssl/rsa/tls.key

---
apiVersion: v1
kind: ConfigMap

metadata:
  name: mailserver.files

data:
  postfix-accounts.cf: |
    xxxxxx|{SHA512-CRYPT}xxxxxx
    xxxxxx|{SHA512-CRYPT}xxxxxx

---
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: dms-pv-claim
  labels:
    app: dms
spec:
  storageClassName: rook-ceph-block
  accessModes:
    - ReadWriteOnce
  resources:
    requests:
      storage: 2Gi

---
apiVersion: v1
kind: Service

metadata:
  name: mailserver
  labels:
    app: mailserver

spec:
  type: LoadBalancer

  selector:
    app: mailserver

  ports:
    # Transfer
    - name: transfer
      port: 25
      targetPort: transfer
      protocol: TCP
    # ESMTP with implicit TLS
    - name: esmtp-implicit
      port: 465
      targetPort: esmtp-implicit
      protocol: TCP
    # ESMTP with explicit TLS (STARTTLS)
    - name: esmtp-explicit
      port: 587
      targetPort: esmtp-explicit
      protocol: TCP
    # IMAPS with implicit TLS
    - name: imap-implicit
      port: 993
      targetPort: imap-implicit
      protocol: TCP

---
apiVersion: apps/v1
kind: Deployment

metadata:
  name: mailserver

  annotations:
    ignore-check.kube-linter.io/run-as-non-root: >-
      'mailserver' needs to run as root
    ignore-check.kube-linter.io/privileged-ports: >-
      'mailserver' needs privilegdes ports
    ignore-check.kube-linter.io/no-read-only-root-fs: >-
      There are too many files written to make The
      root FS read-only

spec:
  replicas: 1
  selector:
    matchLabels:
      app: mailserver

  template:
    metadata:
      labels:
        app: mailserver

      annotations:
        container.apparmor.security.beta.kubernetes.io/mailserver: runtime/default

    spec:
      hostname: smtp
      containers:
        - name: mailserver
          image: ghcr.io/docker-mailserver/docker-mailserver:latest
          imagePullPolicy: IfNotPresent

          securityContext:
            allowPrivilegeEscalation: false
            readOnlyRootFilesystem: false
            runAsUser: 0
            runAsGroup: 0
            runAsNonRoot: false
            privileged: false
            capabilities:
              add:
                # file permission capabilities
                - CHOWN
                - FOWNER
                - MKNOD
                - SETGID
                - SETUID
                - DAC_OVERRIDE
                # network capabilities
                - NET_ADMIN  # needed for F2B
                - NET_RAW    # needed for F2B
                - NET_BIND_SERVICE
                # miscellaneous  capabilities
                - SYS_CHROOT
                - KILL
              drop: [ALL]
            seccompProfile:
              type: RuntimeDefault

          # You want to tune this to your needs. If you disable ClamAV,
          #   you can use less RAM and CPU. This becomes important in
          #   case you're low on resources and Kubernetes refuses to
          #   schedule new pods.
          resources:
            limits:
              memory: 4Gi
              cpu: 1500m
            requests:
              memory: 2Gi
              cpu: 600m

          volumeMounts:
            - name: files
              subPath: postfix-accounts.cf
              mountPath: /tmp/docker-mailserver/postfix-accounts.cf
              readOnly: true

            # PVCs
            - name: dms-pv-claim
              mountPath: /
              subPath: var
              readOnly: false

            # certificates
#            - name: mail-tls-certificate-rsa
#              mountPath: /secrets/ssl/rsa/
#              readOnly: true

            # other
            - name: tmp-files
              mountPath: /tmp
              readOnly: false

          ports:
            - name: transfer
              containerPort: 25
              protocol: TCP
            - name: esmtp-implicit
              containerPort: 465
              protocol: TCP
            - name: esmtp-explicit
              containerPort: 587
            - name: imap-implicit
              containerPort: 993
              protocol: TCP

          envFrom:
            - configMapRef:
                name: mailserver.environment

      restartPolicy: Always

      volumes:
        # configuration files
        - name: files
          configMap:
            name: mailserver.files

        # PVCs
        - name: dms-pv-claim
          persistentVolumeClaim:
            claimName: dms-pv-claim

        # certificates
        - name: certificates-rsa
          secret:
            secretName: mail-tls-certificate-rsa
            items:
              - key: tls.key
                path: tls.key
              - key: tls.crt
                path: tls.crt

        # other
        - name: tmp-files
          emptyDir: {}

---
apiVersion: cert-manager.io/v1
kind: Certificate

metadata:
  name: mail-tls-certificate-rsa

spec:
  secretName: mail-tls-certificate-rsa
  isCA: false
  privateKey:
    algorithm: RSA
    encoding: PKCS1
    size: 2048
  dnsNames: [xxxxxx]
  issuerRef:
    name: selfsigned-issuer
    kind: Issuer

---
apiVersion: v1
kind: ConfigMap
metadata:
  name: tcp-services
  namespace: ingress-nginx
data:
  25:  "mailserver/mailserver:25::PROXY"
  465: "mailserver/mailserver:465::PROXY"
  587: "mailserver/mailserver:587::PROXY"
  993: "mailserver/mailserver:993::PROXY"
```

### 📜 Relevant log output

```Text
0s          Warning   Failed                    pod/mailserver-5f6d96c56f-lrfgm                     (combined from similar events): Error: container create failed: time="2023-06-15T11:30:42+02:00" level=error msg="container_linux.go:380: starting container process caused: process_linux.go:545: container init caused: apply apparmor profile: apparmor failed to apply profile: open /proc/self/attr/exec: no such file or directory"
```

**Comment by georglauterbach** [comment]
> ie: I know that there is a missing crt/key volumeMount (it's revelant?)

This is absolutely relevant! Without it, the requirements of scheduling the pod are not met, and your pod is not scheduled. Not a DMS issue.

**Comment by zelogik** [comment]
@georglauterbach Thanks for your reply!

Was in hope and have mounted a self-signed certificate as configMap.
But always the same problem...


cmd result: kubectl describe pod mailserver-xxxx

```
Name:             mailserver-66c49c4486-nvjsb
Namespace:        default
Priority:         0
Service Account:  default
Node:             worker03/10.140.20.97
Start Time:       Fri, 16 Jun 2023 16:49:35 +0200
Labels:           app=mailserver
                  pod-template-hash=66c49c4486
Annotations:      container.apparmor.security.beta.kubernetes.io/mailserver: runtime/default
Status:           Pending
IP:               10.244.2.251
IPs:
  IP:           10.244.2.251
Controlled By:  ReplicaSet/mailserver-66c49c4486
Containers:
  mailserver:
    Container ID:    
    Image:           ghcr.io/docker-mailserver/docker-mailserver:latest
    Image ID:        
    Ports:           25/TCP, 465/TCP, 587/TCP, 993/TCP
    Host Ports:      0/TCP, 0/TCP, 0/TCP, 0/TCP
    SeccompProfile:  RuntimeDefault
    State:           Waiting
      Reason:        CreateContainerError
    Ready:           False
    Restart Count:   0
    Limits:
      cpu:     1500m
      memory:  4Gi
    Requests:
      cpu:     600m
      memory:  2Gi
    Environment Variables from:
      mailserver.environment  ConfigMap  Optional: false
    Environment:              <none>
    Mounts:
      / from dms-pv-claim (rw,path="var")
      /secrets/ssl/rsa/ from my-tls-secret (ro)
      /tmp from tmp-files (rw)
      /tmp/docker-mailserver/postfix-accounts.cf from files (ro,path="postfix-accounts.cf")
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-qlxkw (ro)
Conditions:
  Type              Status
  Initialized       True 
  Ready             False 
  ContainersReady   False 
  PodScheduled      True 
Volumes:
  files:
    Type:      ConfigMap (a volume populated by a ConfigMap)
    Name:      mailserver.files
    Optional:  false
  dms-pv-claim:
    Type:       PersistentVolumeClaim (a reference to a PersistentVolumeClaim in the same namespace)
    ClaimName:  dms-pv-claim
    ReadOnly:   false
  my-tls-secret:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  my-tls-secret
    Optional:    false
  tmp-files:
    Type:       EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium:     
    SizeLimit:  <unset>
  kube-api-access-qlxkw:
    Type:                    Projected (a volume that contains injected data from multiple sources)
    TokenExpirationSeconds:  3607
    ConfigMapName:           kube-root-ca.crt
    ConfigMapOptional:       <nil>
    DownwardAPI:             true
QoS Class:                   Burstable
Node-Selectors:              <none>
Tolerations:                 node.kubernetes.io/not-ready:NoExecute op=Exists for 300s
                             node.kubernetes.io/unreachable:NoExecute op=Exists for 300s
Events:
  Type     Reason                  Age                From                     Message
  ----     ------                  ----               ----                     -------
  Warning  FailedScheduling        46s (x2 over 47s)  default-scheduler        0/4 nodes are available: pod has unbound immediate PersistentVolumeClaims. preemption: 0/4 nodes are available: 4 No preemption victims found for incoming pod..
  Normal   Scheduled               43s                default-scheduler        Successfully assigned default/mailserver-66c49c4486-nvjsb to worker03
  Normal   SuccessfulAttachVolume  42s                attachdetach-controller  AttachVolume.Attach succeeded for volume "pvc-474a7feb-d5f6-4272-ba39-1ef2145e57d4"
  Warning  Failed                  40s                kubelet                  Error: container create failed: time="2023-06-16T16:49:38+02:00" level=error msg="container_linux.go:380: starting container process caused: process_linux.go:545: container init caused: apply apparmor profile: apparmor failed to apply profile: open /proc/self/attr/exec: no such file or directory"
  Warning  Failed                  38s                kubelet                  Error: container create failed: time="2023-06-16T16:49:39+02:00" level=error msg="container_linux.go:380: starting container process caused: process_linux.go:545: container init caused: apply apparmor profile: apparmor failed to apply profile: open /proc/self/attr/exec: no such file or directory"
  Warning  Failed                  27s                kubelet                  Error: container create failed: time="2023-06-16T16:49:51+02:00" level=error msg="container_linux.go:380: starting container process caused: process_linux.go:545: container init caused: apply apparmor profile: apparmor failed to apply profile: open /proc/self/attr/exec: no such file or directory"
  Warning  Failed                  12s                kubelet                  Error: container create failed: time="2023-06-16T16:50:06+02:00" level=error msg="container_linux.go:380: starting container process caused: process_linux.go:545: container init caused: apply apparmor profile: apparmor failed to apply profile: open /proc/self/attr/exec: no such file or directory"
  Normal   Pulled                  2s (x5 over 40s)   kubelet                  Container image "ghcr.io/docker-mailserver/docker-mailserver:latest" already present on machine
  Warning  Failed                  1s                 kubelet                  Error: container create failed: time="2023-06-16T16:50:17+02:00" level=error msg="container_linux.go:380: starting container process caused: process_linux.go:545: container init caused: apply apparmor profile: apparmor failed to apply profile: open /proc/self/attr/exec: no such file or directory"

```
So /secrets/ssl/rsa/, and the only error I got is really from that apparmor... even if 
```Annotations:      container.apparmor.security.beta.kubernetes.io/mailserver: runtime/default```
is applied (tested with unconfined too)

**Comment by georglauterbach** [comment]
No, there first error is this one, and you need to take care of it:

```txt
Warning  FailedScheduling        46s (x2 over 47s)  default-scheduler        0/4 nodes are available: pod has unbound immediate PersistentVolumeClaims. preemption: 0/4 nodes are available: 4 No preemption victims found for incoming pod..
```

**Comment by zelogik** [comment]
It's "normal" it's a rook/ceph volume, on a low power cluster for test. But the line below that error is:
```
Normal   Scheduled               43s                default-scheduler        Successfully assigned default/mailserver-66c49c4486-nvjsb to worker03
```


I have tested with a local pvc and the same problem.... ( just remove the warning with the pvc)
```Events:
  Type     Reason     Age                    From               Message
  ----     ------     ----                   ----               -------
  Normal   Scheduled  4m14s                  default-scheduler  Successfully assigned default/mailserver-84848965c5-8ghgq to worker03
  Warning  Failed     4m11s (x2 over 4m12s)  kubelet            Error: container create failed: time="2023-06-19T08:40:50+02:00" level=error msg="container_linux.go:380: starting container process caused: process_linux.go:545: container init caused: apply apparmor profile: apparmor failed to apply profile: open /proc/self/attr/exec: no such file or directory"
  Warning  Failed     3m56s                  kubelet            Error: container create failed: time="2023-06-19T08:41:06+02:00" level=error msg="container_linux.go:380: starting container process caused: process_linux.go:545: container init caused: apply apparmor profile: apparmor failed to apply profile: open /proc/self/attr/exec: no such file or directory"
  Warning  Failed     3m42s                  kubelet            Error: container create failed: time="2023-06-19T08:41:20+02:00" level=error msg="container_linux.go:380: starting container process caused: process_linux.go:545: container init caused: apply apparmor profile: apparmor failed to apply profile: open /proc/self/attr/exec: no such file or directory"
  Warning  Failed     3m30s                  kubelet            Error: container create failed: time="2023-06-19T08:41:32+02:00" level=error msg="container_linux.go:380: starting container process caused: process_linux.go:545: container init caused: apply apparmor profile: apparmor failed to apply profile: open /proc/self/attr/exec: no such file or directory"
```

**Comment by georglauterbach** [comment]
Then remove apparmor completely from the deployment. It's a nice-to-have when starting out with DMS. Moreover, this is a L8s thing, not DMS, and we're in the DMS issue tracker, not in the K8s.. I'd therefore like to close this. Is there anything related to DMS specifically here left?

**Comment by zelogik** [comment]
I know that it's could be more a k8s problem than a DMS issue, but I have many deployment on that cluster that use apparmor without any issues... it´s why at first I opened a issue there and not on k8s :-D

Regards

**Comment by zelogik** [comment]
I close, even if the issue is not resolved... 
I can't confirm that the problem is DMS related, even if it's the only deployment that have that issue with apparmor profile on my k8s cluster.

thanks @georglauterbach

**Comment by georglauterbach** [comment]
Mmmhh, I see - very weird indeed. I'd remove everything and start afresh. The apple everything one by one. For me, this has always been the best strategy debugging such issues. Also, try to use the slimmest possible deployment to begin with. Good luck 🚀
